website/blog/deployers.md

```diff
@@ -0,0 +1,525 @@
+exports logs and traces to [Cloud Logging][cloud_logging] and [Cloud
+Trace][cloud_trace].
+
+### Listeners
+
+Next, we implement the listener methods. When a component requests a network
+listener, the `GetAddress` method is invoked. This method returns the address on
+which the component should listen. Our simple deployer always returns
+`"localhost:0"`. After the requesting weavelet receives this address, it
+creates a network listener on the address and invokes the `ExportListener`
+method with the concrete address that it's listening on. For example,
+`"localhost:0"` could be resolved to `"127.0.0.1:35879"`.
+
+```go
+// Responsibility 2: Listeners.
+func (h *handler) GetAddress(req *protos.GetAddressRequest) (*protos.GetAddressReply, error) {
+    return &protos.GetAddressReply{Address: "localhost:0"}, nil
+}
+
+func (h *handler) ExportListener(req *protos.ExportListenerRequest) (*protos.ExportListenerReply, error) {
+    return &protos.ExportListenerReply{}, nil
+}
+```
+
+`weaver multi`'s implementation of `GetAddress` always returns `"localhost:0"`.
+Its `ExportListener` runs a local HTTP proxy on the address specified in the
+`LocalAddress` field of the [ListenerOptions][] passed to
+[Instance.Listener][Listener]. This proxy balances traffic across the addresses
+of the listeners reported to `ExportListener`. This balances traffic across all
+replicas of a listener. `weaver gke` implements listeners using load balancers
+in Google Cloud.
+
+### Components
+
+Next, we implement `StartComponent`. When a component hosted by a weavelet calls
+`weaver.Get[T]` for some component `T`, the weavelet calls the `StartComponent`
+method for component `T`. `StartComponent` should start the
+component&mdash;potentially with multiple replicas&mdash;if it hasn't already
+been started.
+
+Our deployer calls `deployer.spawn` to spawn the weavelet hosting the requested
+component, it it hasn't already been spawned. The deployer then calls
+`UpdateRoutingInfo` to inform the weavelet of the address of the component it
+requested. This allows the weavelet to perform RPCs with the component.
+
+A deployer should call `UpdateRoutingInfo` whenever there is a change to the
+routing information of a component that a weavelet has called `StartComponent`
+on. For example, if a deployer detects that a weavelet hosting component `A` has
+crashed, it should call `UpdateRoutingInfo` on all weavelets that have called
+`StartComponent` on `A` with new routing information that omits the address of
+the failed weavelet.
+
+```go
+// Responsibility 3: Components.
+func (h *handler) StartComponent(req *protos.StartComponentRequest) error {
+    w, err := h.deployer.spawn(req.Component)
+    if err != nil {
+        return err
+    }
+    h.envelope.UpdateRoutingInfo(&protos.UpdateRoutingInfo{Replicas: []string{w.address}})
+    return nil
+}
+```
+
+Note that if a weavelet should perform component method calls locally, a
```

Comment:
This paragraph may be potentially confusing.

See if you can come up with something clearer?